### ISSUE TITLE
Tensor Dot for rank bigger than 3

### ISSUE THREAD

**Issue by aidevnn** [SciSharp/NumSharp#350]
Hello

I see Tensor Dot for rank bigger than 3 are not implemented but it can be done with ease.

Tensor Dot Tensor is equivalent to a Matrix Dot Matrix after reshape and transpose then reshape

for example, in python numpy
```
a=np.random.randint(0,10,(4,2,3))
b=np.random.randint(0,10,(5,3,6))

a.dot(b) 
a.reshape(-1,3).dot(b.transpose(1,0,2).reshape(3,-1)).reshape(4,2,5,6)
```

and be generalized for every rank.
A Tensor Dot for rank bigger than 3 is not used in common DeepLearning layers, but it can be done with no cost.

Tx

**Comment by Nucs** [comment]
What version of NumSharp are you using? The newest supports up to 16-16 dimensions.

**Comment by aidevnn** [comment]
@Nucs 
i have take a look at this line
https://github.com/SciSharp/NumSharp/blob/master/src/NumSharp.Core/Backends/Default/Math/BLAS/Default.Dot.cs#L75

Ok, if the tensor dot for hger rank are already implemented.

**Comment by Nucs** [comment]
If you are getting `NotSuppotedException` then you are not using `np.dot` in a mathematically proper way. The implementation is strictly equivalent to what numpy allows to do.
[Heres](https://raw.githubusercontent.com/SciSharp/NumSharp/master/src/NumSharp.Core/Backends/Default/Math/BLAS/Default.Dot.NDMD.cs) the nd-md implementation that is called right before the throw.

**Comment by Nucs** [comment]
The following code works.
```C#
            var a = np.random.randint(0, 10, (4, 2, 3));
            var b = np.random.randint(0, 10, (4, 3, 2));

            a = np.dot(a, b);
            Console.WriteLine((string) a);
            Console.ReadLine();
```
As mentioned, transpose is yet to be implemented so your full code can be used yet.

**Comment by aidevnn** [comment]
I understand recently that high order Tensor can be rewrited as a batch of common Matrix.
Thank you for the C# example.
Sorry for this useless issue.

**Comment by Nucs** [comment]
No worries, better ask than sit in silence.
Related to https://github.com/SciSharp/NumSharp/issues/117
Closed.

**Comment by Nucs** [comment]
`np.swapaxes`, `np.transpose`, `np.moveaxis` and `np.rollaxis` were added in the following [this commit](https://github.com/SciSharp/NumSharp/commit/39896599ac40584821802cfd63a15458060289ba ).
And will be released soon in 0.20.4.
